### PR TITLE
feat: implement KECCAK256 opcode

### LINKED ISSUE #29: Opcode: KECCAK256
We'll need to use https://crates.io/crates/sha3 for this

### PR DESCRIPTION
**Pull request by Danikim01**
closes #29 
This PR implements the KEKKACK256 opcode

### CODE REVIEW

File: tests/operations.rs [hunk 1]
```diff
@@ -102,6 +103,28 @@ pub fn biguint_256_from_bigint(value: BigInt) -> BigUint {
     }
 }
 
+#[test]
```
Comment (JulianVentura) [inline tests/operations.rs:106]: Can we add two more tests? One with the [EVM Codes Keccak256 playground example](https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code=%27sPutkrequired%20valugin%20memoryj32%200xFFFFFFFFffffz0wMSTOREwwsCallkopcodez4z0wKECCAK256%27~0000000zj1%20w%5Cns%2F%2F%20k%20thgjwPUSHge%20f~~%01fgjkswz~_) and the other one with a stack overflow?

Comment (Danikim01) [reply tests/operations.rs:106]: resolved, i also added a test for underflow.

File: src/codegen/operations.rs [hunk 2]
```diff
@@ -159,6 +161,110 @@ fn codegen_callvalue<'c, 'r>(
...
+fn codegen_keccak256<'c, 'r>(
+    op_ctx: &mut OperationCtx<'c>,
+    region: &'r Region<'c>,
+) -> Result<(BlockRef<'c, 'r>, BlockRef<'c, 'r>), CodegenError> {
+    let start_block = region.append_block(Block::new(&[]));
+    let context = &op_ctx.mlir_context;
+    let location = Location::unknown(context);
+
+    let stack_flag = check_stack_has_space_for(context, &start_block, 1)?;
+    let flag = check_stack_has_at_least(context, &start_block, 2)?;
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:173]: We only need to check the stack has at least two elements here. Popping those two elements will make space for the result.

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -8,10 +8,12 @@ license = "LICENSE"
 keywords = ["ethereum", "evm"]
 
 [dependencies]
+hex-literal = "0.4.1"
```
[the remark this answers is not in the document]
Comment (Danikim01) [reply Cargo.toml]: resolved

File: src/codegen/operations.rs (hunk 2 above)
Comment (Danikim01) [reply src/codegen/operations.rs]: resolved

File: src/codegen/context.rs [hunk 4]
```diff
@@ -463,6 +463,25 @@ impl<'c> OperationCtx<'c> {
         )
     }
 
+    pub(crate) fn keccak256_hasher_syscall(
```
[the remark this answers is not in the document]
Comment (Danikim01) [reply src/codegen/context.rs]: resolved

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**